ui/src/app/views/workspace-edit/component.ts

```diff
@@ -122,31 +123,65 @@ export class WorkspaceEditComponent implements OnInit {
     this.savingWorkspace = false;
   }
 
+  validateForm(): boolean {
+    if (this.savingWorkspace) {
+      return false;
+    }
+    this.nameNotEntered = isBlank(this.workspace.name);
+    if (this.nameNotEntered) {
+      return false;
+    }
+    return true;
+  }
+
+  addWorkspace(): void {
+    if (!this.validateForm()) {
+      return;
+    }
+    this.savingWorkspace = true;
+    this.errorHandlingService.retryApi(
+      this.workspacesService.createWorkspace(this.workspace))
+      .subscribe(
+        () => this.navigateBack(),
+        (error) => this.workspaceCreationError = true);
+  }
+
   updateWorkspace(): void {
-    if (!this.savingWorkspace) {
-      if (isBlank(this.workspace.name)) {
-        this.nameNotEntered = true;
-      } else {
-        this.savingWorkspace = true;
-        this.nameNotEntered = false;
-        this.errorHandlingService.retryApi(this.workspacesService.updateWorkspace(
-            this.oldWorkspaceNamespace,
-            this.oldWorkspaceName,
-            this.workspace))
-          .subscribe(
-            () => {
-              this.navigateBack();
-            },
-            (error) => {
-              if (error.status === 409) {
-                this.workspaceUpdateConflictError = true;
-              } else {
-                this.workspaceUpdateError = true;
-              }
-            });
-      }
+    if (!this.validateForm()) {
+      return;
     }
+    this.savingWorkspace = true;
+    this.errorHandlingService.retryApi(this.workspacesService.updateWorkspace(
+      this.oldWorkspaceNamespace,
+      this.oldWorkspaceName,
+      this.workspace))
+      .subscribe(
+        () => this.navigateBack(),
+        (error) => {
+          if (error.status === 409) {
+            this.workspaceUpdateConflictError = true;
+          } else {
+            this.workspaceUpdateError = true;
+          }
+        });
   }
+
+  cloneWorkspace(): void {
+    if (!this.validateForm()) {
+      return;
+    }
+    this.savingWorkspace = true;
+    this.errorHandlingService.retryApi(this.workspacesService.cloneWorkspace(
+      this.oldWorkspaceNamespace,
+      this.oldWorkspaceName, {
+        workspace: this.workspace,
```

Comment:
Is there a reason to have updateWorkspace and cloneWorkspace take different argument types? (I think I prefer the passing of workspace as a new object, I just wanted to check on consistency. If the answer is: clone might have more fields added to it, that's fine in my book.)

Reply:
Yes, that is the reason. Clone could easily need additional parameters that we would not want to encode into the Workspace model. Update should be doing the same (and I would propose that we change it to pass an UpdateWorkspaceRequest instead) e.g. see the GCP style guide for a REST update method: https://cloud.google.com/apis/design/standard_methods#update

In general when defining an API, it's best to avoid likely situations where you think you'd need to break the API to expand it, so as a rule of thumb I'd avoid directly embedding a proper resource type as the sole request/response outside of standard REST methods. Of course there are other outs with Swagger, e.g. query params, but overuse of these escape hatches can will make for an inconsistent API.

Reply:
Sounds good to me. I agree that we should probably move to UpdateWorkspaceRequest. Do you want to do a follow up on that, or should I work on that while you wrap up clone?

Reply:
Sure, thanks - feel free to send me a PR. If you don't have time for it, I can also likely get around to it later today or next week.

Reply:
Yeah, I will work on getting that out next. (Just got admin CLI to PR ready, so I can do a couple of cleanup PRs next.)